tests/returntypes/test_autograd_new.py

```diff
@@ -0,0 +1,1415 @@
+            @classmethod
+            def capabilities(cls):
+                capabilities = super().capabilities().copy()
+                capabilities.update(
+                    provides_jacobian=True,
+                )
+                return capabilities
+
+            def jacobian(self, *args, **kwargs):
+                raise NotImplementedError()
+
+        dev = MyQubit(wires=2)
+
+        @qml.qnode(dev, diff_method="parameter-shift", mode="backward")
+        def qnode(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))
+
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=True)
+
+        def cost(a, b):
+            return autograd.numpy.hstack(qnode(a, b))
+
+        res = qml.jacobian(cost)(a, b)
+        expected = ([-np.sin(a), np.sin(a) * np.sin(b)], [0, -np.cos(a) * np.cos(b)])
+
+        assert np.allclose(res[0], expected[0])
+        assert np.allclose(res[1], expected[1])
+
+
+class SpecialObject:
+    """SpecialObject
+    A special object that conveniently encapsulates the return value of
+    a special observable supported by a special device and which supports
+    multiplication with scalars and addition.
+    """
+
+    def __init__(self, val):
+        self.val = val
+
+    def __mul__(self, other):
+        new = SpecialObject(self.val)
+        new *= other
+        return new
+
+    def __imul__(self, other):
+        self.val *= other
+        return self
+
+    def __rmul__(self, other):
+        return self * other
+
+    def __iadd__(self, other):
+        self.val += other.val if isinstance(other, self.__class__) else other
+        return self
+
+    def __add__(self, other):
+        new = SpecialObject(self.val)
+        new += other.val if isinstance(other, self.__class__) else other
+        return new
+
+    def __radd__(self, other):
+        return self + other
+
+
+class SpecialObservable(Observable):
+    """SpecialObservable"""
+
+    num_wires = AnyWires
+    num_params = 0
+    par_domain = None
+
+    def diagonalizing_gates(self):
+        """Diagonalizing gates"""
+        return []
+
+
+class DeviceSupportingSpecialObservable(DefaultQubit):
+    name = "Device supporting SpecialObservable"
+    short_name = "default.qubit.specialobservable"
+    observables = DefaultQubit.observables.union({"SpecialObservable"})
+
+    @staticmethod
+    def _asarray(arr, dtype=None):
+        return arr
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            provides_jacobian=True,
+        )
+        return capabilities
+
+    def expval(self, observable, **kwargs):
+        if self.analytic and isinstance(observable, SpecialObservable):
+            val = super().expval(qml.PauliZ(wires=0), **kwargs)
+            return np.array(SpecialObject(val))
```

Comment:
Casting to an array here is a major change from before, it will have to be communicated to users well.

Reply:
Yes, any idea where?